src/codegen/constant-propagation.lisp

```diff
@@ -0,0 +1,182 @@
+(defpackage #:coalton-impl/codegen/constant-propagation
+  (:use
+   #:cl
+   #:coalton-impl/codegen/ast)
+  (:local-nicknames
+   (#:tc #:coalton-impl/typechecker))
+  (:export
+   #:propagate-constants))
+
+(in-package #:coalton-impl/codegen/constant-propagation)
+
+(defun constant-var-value (var constant-bindings)
+  (cdr (assoc var constant-bindings)))
```

Comment:
when checking that value exists later you need the whole cell, not just its `cdr`, which might still exist but be nil.
you can do this similar to `gethash`:
```suggestion
  (let ((x (assoc var constant-bindings)))
    (if (null x)
        (values nil nil) ; not found
        (values x t))))
```
(and then change usage of this function accordingly)

Reply:
if only there were a type system (: